### PR TITLE
Subdirectory

### LINKED ISSUE #147: Subfolder support (publicPath)
Problem:
I'm using nginx to proxy **localhost:3001** to **domain.com/kuma**, but at the moment vue app expects that everything is served from root (domain.com/assets/vendor.a54d5402.js) and all assets get 404, because all files are in kuma subfolder (domain.com/kuma/assets/vendor.a54d5402.js) not in root directory.

Solution:
I'm not familiar with vue, but I guess it needs this public path setting for relative paths:
https://cli.vuejs.org/config/#publicpath
https://vitejs.dev/config/#base

If there are no other obvious limitations then please consider using relative paths.

It is great app & thanks for your work!

### PR DESCRIPTION
**Pull request by trogper**
# Description

This adds support for Kuma deployment in http path subfolder. I have added two environment variables, `UPTIME_KUMA_BASE_PATH` and `BASE_PATH`. The value should start and end with `/`, but there is sanitization, which will add them.
I have changed relevant paths in vue and html from absolute to relative and added html base tag, which is used byt the browser to build absolute paths from relative. Then I made express mount the former root to the specified `BASE_PATH`.  Additionally I have fixed a few `<a>` tags in vie to proper `<router-link>`
Socket.io does not support base tag by itself (it builds the connection url manually), so we have to prepend the basePath manually.

Fixes #147 

## TODO
There are a few things I would like someone to help me with:
1. where is the proper place to put `basePath` calculation/variable? Currently it is in `src\mixins\socket.js:71` and `src\components\settings\General.vue:186`
2. create tests for it
3. test it manually on their server/deployment (I have deployed it on mine already)

## Type of change
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and test it
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

### CODE REVIEW

File: src/mixins/socket.js [hunk 1]
```diff
@@ -67,8 +67,12 @@ export default {
                 wsHost = protocol + location.host;
             }
 
+            // always starts and ends with '/'
+            const basePath = document.querySelector("head base").getAttribute("href");
+
             socket = io(wsHost, {
                 transports: ["websocket"],
+                path: basePath + "socket.io"
```
Comment (Saibamen) [inline src/mixins/socket.js:75]:
```suggestion
                path: basePath + "socket.io",
```

**Review by Saibamen — CHANGES_REQUESTED** [review]

**Comment by trogper** [comment]
@louislam can you please help me with the implementation details? I have a few questions in TODO in the first comment.

**Comment by ckocyigit** [comment]
I have merged this PR in my fork and experienced a strange behaviour.

When pressind the Edit Button on an already existing Notification it redirects me to the Status page. 

I wanted to open a Issue to Kuma directly when it occurred to me that I have this probably "not fully tested" PR in my fork. 

Maybe @trogper can you check if this happens on your site too?

**Comment by zacharlie** [comment]
For anyone interested I have created a dockerfile that implements this, which I assume can be used as a temporary measure until there is some for of resolution for #1349

https://github.com/zacharlie/spatial-dbstack/blob/main/uptime-kuma-dockerfile

**Comment by trogper** [comment]
Yes, the are minor conflicts. I'll solve them when I get some time for it

**Comment by zacharlie** [comment]
@trogper just a heads up I also tested and noticed the issue described above with notification editing when merging this

**Comment by trogper** [comment]
Thanks, I know about the issue already

**Comment by pm-pm** [comment]
I am glad to see that progress is being made. 
Thank you

**Comment by trogper** [comment]
@pm-pm @zacharlie 
Conflicts resolved, should be mergable now.

**Comment by pm-pm** [comment]
Push this heavy topic. ;-)

File: src/mixins/socket.js [hunk 2]
```diff
@@ -96,8 +96,12 @@ export default {
                 wsHost = protocol + location.host;
             }
 
+            // always starts and ends with '/'
+            const basePath = document.querySelector("head base").getAttribute("href");
+
             socket = io(wsHost, {
-                transports: [ "websocket" ],
+                transports: ["websocket"],
```
Comment (Saibamen) [inline src/mixins/socket.js:103]:
```suggestion
                transports: [ "websocket" ],
```

Comment (trogper) [reply src/mixins/socket.js:103]: I wonder how that happened, since I prefer the original style...

Comment (Saibamen) [reply src/mixins/socket.js:103]: We are using ESLint: https://github.com/louislam/uptime-kuma/blob/master/.eslintrc.js

Comment (trogper) [reply src/mixins/socket.js:103]: I mean I wonder how did the change from spaced to non-spaced happen, since I would not do it myself. I would leave it spaced.

**Review by Computroniks — APPROVED** [review]

**Review by BioSs54 — APPROVED** [review]

**Review by d1ceward — APPROVED** [review]

**Comment by louislam** [comment]
I put this into to 2.0 milestone, so I don't need to worry about backward compatibility.

**Comment by trogper** [comment]
Is there any reason why this would not be backwards compatible?
Do you have any expectation when will the 2.0 be released?

**Comment by trogper** [comment]
I understand that external links are `<a>`, but why should some internal be too?

**Comment by jonaslb** [comment]
> For this pr, I think it is hard to predict whether it will cause any side effects to current users, as it is changed a lot of base code. 

I can't imagine how this change would have averse side effects. All the old paths still work, right? For this reason it seems very appropriate even for a minor point release.

If you want to be really careful, maybe you could merge it into a 1.20.0beta1 release before a "full" release? If there are problems it will almost surely surface then. And then the decision can be to either do a bugfix before full release or pull it back.

**Comment by EugeneDae** [comment]
Looking forward to this, thanks to everyone who works on this feature. That said, its a pity that modern fancy JS frameworks and libs require so much effort from programmers to support such a simple and necessary setting.

**Comment by Ex0danify** [comment]
I tried this PR on my local machine, I can't seem to get it to work.
The ENV var UPTIME_KUMA_BASE_PATH=/kuma/ is set, I proxy uptime-kuma to /kuma via apache reverse proxy, but all I get when visiting my localhost is the following: Cannot GET /

Am I using this wrong?

**Comment by Ex0danify** [comment]
Sure, sry for not posting it the first time. It really is jsut for testing it with a subdirectory on my local Linux machine, so pardon me if the config is not hardened etc.

```
VirtualHost *:80>
ServerAdmin admin@localhost.com
DocumentRoot "/var/www/html"
ErrorLog "/var/log/httpd/localhost-error_log"
CustomLog "/var/log/httpd/localhost-access_log" common

<Proxy *>
    Require all granted
</Proxy>

ProxyRequests     Off
ProxyPreserveHost On
ProxyTimeout      2000

ProxyPass /kuma http://localhost:8081/
RewriteEngine on
RewriteCond %{HTTP:Upgrade} websocket [NC]
RewriteCond %{HTTP:Connection} upgrade [NC]
RewriteRule ^/kuma/?(.*) "ws://localhost:8081/$1" [P,L]

</VirtualHost>

**Comment by trogper** [comment]
One of the issues is that you proxypass/rewrite `/kuma` (no trailing slash) to `:8081/` (with), which means `/kuma/` goes to `:8081//`
Actually my PR makes kuma also host on `/kuma`, so you need to proxypass/rewite to `:8081/kuma`. I don't recommend rewriting with trailing slashes, because requests to root (`/kuma` without trl.slash) won't get proxied.

This is my config:
```
RewriteEngine On
ProxyPass /kuma http://127.0.0.1:3018/kuma
RewriteEngine on
RewriteCond %{HTTP:Upgrade} websocket [NC]
RewriteCond %{HTTP:Connection} upgrade [NC]
RewriteRule ^/kuma/(.*) "ws://127.0.0.1:3018/kuma/$1" [P,L]
```

**Comment by Ex0danify** [comment]
> One of the issues is that you proxypass/rewrite `/kuma` (no trailing slash) to `:8081/` (with), which means `/kuma/` goes to `:8081//` Actually my PR makes kuma also host on `/kuma`, so you need to proxypass/rewite to `:8081/kuma`. I don't recommend rewriting with trailing slashes, because requests to root (`/kuma` without trl.slash) won't get proxied.
> 
> This is my config:
> 
> ```
> RewriteEngine On
> ProxyPass /kuma http://127.0.0.1:3018/kuma
> RewriteEngine on
> RewriteCond %{HTTP:Upgrade} websocket [NC]
> RewriteCond %{HTTP:Connection} upgrade [NC]
> RewriteRule ^/kuma/(.*) "ws://127.0.0.1:3018/kuma/$1" [P,L]
> ```

Thanks for your reply!
I tried modifying my proxy config with your example, but sadly, nothing really changed:
![image](https://github.com/louislam/uptime-kuma/assets/51802014/50eca733-9d38-47ab-8799-56e6916553c7)


This is my new apache config:
```
<VirtualHost *:80>
    ServerAdmin admin@localhost.com
    DocumentRoot "/var/www/html"
    ErrorLog "/var/log/httpd/localhost-error_log"
    CustomLog "/var/log/httpd/localhost-access_log" common

    <Proxy *>
        Require all granted
    </Proxy>

ProxyRequests     Off
ProxyPreserveHost On
ProxyTimeout      2000

ProxyPass /kuma http://localhost:8081/kuma
RewriteEngine on
RewriteCond %{HTTP:Upgrade} websocket [NC]
RewriteCond %{HTTP:Connection} upgrade [NC]
RewriteRule ^/kuma/(.*) "ws://localhost:8081/kuma/$1" [P,L]

</VirtualHost>
```
Made double sure, that the env var is respected by the terminal and the server itself:
```
$> echo $UPTIME_KUMA_BASE_PATH                                                                                                                                                                                                                                                                                           
/kuma/
```

![image](https://github.com/louislam/uptime-kuma/assets/51802014/4d7bc3b5-a67e-42c2-ab3b-7650f11428a8)

This is how it looks when I try to access the dashboard:
![image](https://github.com/louislam/uptime-kuma/assets/51802014/bda06718-050a-423f-ad0f-cc5e3dcff46a)

This is the message when I access uptimekuma without the trailing slash in the url:
![image](https://github.com/louislam/uptime-kuma/assets/51802014/09925661-55f7-4957-b099-574617aa54aa)


I also tried another proxy url for the uptimekuma instance:
```
<VirtualHost *:80>
    ServerAdmin admin@localhost.com
    DocumentRoot "/var/www/html"
    ErrorLog "/var/log/httpd/localhost-error_log"
    CustomLog "/var/log/httpd/localhost-access_log" common

    <Proxy *>
        Require all granted
    </Proxy>

ProxyRequests     Off
ProxyPreserveHost On
ProxyTimeout      2000

ProxyPass /admin/uptime http://localhost:8081/kuma
RewriteEngine on
RewriteCond %{HTTP:Upgrade} websocket [NC]
RewriteCond %{HTTP:Connection} upgrade [NC]
RewriteRule ^/admin/uptime/(.*) "ws://localhost:8081/kuma/$1" [P,L]

</VirtualHost>
```

The result:
![image](https://github.com/louislam/uptime-kuma/assets/51802014/2db29e1b-4a75-4523-83d1-654ed39d1c7a)


What really is strange to me, is that I can't even reach the web interface on localhost directly:
![image](https://github.com/louislam/uptime-kuma/assets/51802014/267600c9-a68e-44f6-a3a6-45b9927af1b8)
![image](https://github.com/louislam/uptime-kuma/assets/51802014/de357fa8-531c-45e5-870a-82832ef3b9cc)

So it seems the proxy is working fine, if that literally is, what the node server responds. I also can't see anything happening in the node server logs, when running with *npm run dev*, when I access the URL.

**Comment by byeong0** [comment]
@louislam 
This PR has gone to 2.0, but we still want it to happen quickly. When is the 2.0 release coming?

**Comment by CommanderStorm** [comment]
@byeong0 
Please refrain from posting `+1` / requests for update-things on PRs, as this makes reviewing it even harder.
PRs are for noting problems and working on them, while for updates, there is a subscribe to this PR button at the top.
[We use 👍🏻 on PRs to prioritise work](https://github.com/louislam/uptime-kuma/pulls?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by CommanderStorm** [comment]
@trogper Do you think integrating https://github.com/louislam/uptime-kuma/issues/3316 would be a good idea?

**Comment by trogper** [comment]
@CommanderStorm 
I'm not quite sure about this, it seems hacky.
When exactly would kuma decide which subdirectory/subpath to use? Probably on the first request.
Should it be persisted across restarts, or should it "listen" on every start?
If on every start, it could be abusable when not behind a proxy, e.g. you would reboot the server, kuma would start automatically and then "Bad person™" would send a request with faked proxy-header, thus your kuma would think it is hosted on subdirectory and would not work as expected.

In conclusion, it would be nice, if it could be done properly or some standardised way.

**Comment by CommanderStorm** [comment]
True, I don't think anymore that integrating this would be a good idea.
Could you add a `Closing #3316` to the description? (Reasoning: This PR makes said issue irrelevant and something we should not integrate)

**Comment by louislam** [comment]
I still lack confidence to merge it, even though in 2.0.0.

- Some `location.href` is intented to be used, it should not change to `vue router`
- The push url is incorrect
image
- Some button or link navigate to a blank page after clicked.

Since there is no full e2e test for Uptime Kuma, it is honestly not easy to be tested. I pull it over of the 2.0.0 milestone.

**Comment by trogper** [comment]
The PR is based on old version, it is expected to not work well with current version, it needs to be updated.

Push URL should be a trivial fix, other bugs probably too. If you tell me you want to have this feature, I will try to merge 2.0 into this PR.

Regarding testing: it is not that big of a project, it should be sufficient to test manually an then rely on beta testers to report bugs.

**Comment by arthegost** [comment]
Wanted to see if there is any work going into this, or is it a dead end?

**Comment by CommanderStorm** [comment]
This is a massive change and, as mentioned above by Louis, currently does not fully work.

Once we have
- better test coverage and
- the bugs which remain are fixed

=> the certainty that this does not inadvertently breaks a lot of sub-features, we can merge this.
Currently the testing effort that has gone into this PR is not enough to bring this to a mergable state.
PRs can either be tested via [this way](https://github.com/louislam/uptime-kuma/wiki/Test-Pull-Requests) or via [writing e2e-testcases here](https://github.com/louislam/uptime-kuma/tree/master/test).
If you/somebody here wants to contribute the second kind, that would be very much appreciated.

The reason Louis is so hesitant about merging, is that shipping broken software is an incredible burden on maintainers and increases the needed support load by a LOT.

If there is certainty that this does not brick some subfeature, I don't see a reason, where this would need a major release.

**Comment by Ionys320** [comment]
I'm not sure if the PR is correctly handling the situation.
About the server/backend, the reverse proxy config is normally enough to avoid needing any change regarding routes management.

Regarding the frontend, it should be enough for most cases to use the `<base>` tag. The best solution would of course build the frontend with the `BASE_PATH` setting, but this is not an option ig. I'll check a way to edit <base>  and avoid needing a lot of changes.

**Comment by Ionys320** [comment]
Question to @CommanderStorm or @louislam: could we run a `npm run build` after Docker pull process? Let me explain: for the moment, we don't really use Vite base path as it is supposed to work. Therefore, we lost a lot of functionalities who would simplify a lot this PR. By being capable to run `npm run buid` after Docker pull, we could build the frontend using Vite by providing it the right `base` path, based on the ENV variable `BASE_PATH`.

I'm checking on my side, and it seems I only need to change a few files (including the times I change `a` tag to `router-link` tag to let Vite take care of it, and adding `import.meta.env.BASE_URL`), it appears I woudln't need to change more than 15 files. I of course need to do more tests and I can optimize some stuff, but I think this solution would be more maintainable than the one @trogper is working on, mainly because it gets around the problem without using the tools normally used for (Not a critic @trogper ofc, your work is awesome regarding the contrainst you got).

**Comment by louislam** [comment]
> could we run a npm run build after Docker pull process

Cannot, unless the docker image includes all dev dependencies / build tools which probably is more complicated and I don't want to.

File: src/components/PingChart.vue
Comment (Nikolas-Lehto) [inline src/components/PingChart.vue:1]: Could you elaborate on these changes? I was reading the diffs to try to understand them with my _very_ limited frontend skills and didn't understand how these are related to subdirectories in any way.

**Comment by mskyttner** [comment]
Sorry for jumping in, great tool, thanks! I just wanted to share that the first thing I wanted to do when learning about and starting to deploy kuma was to deploy the app with the context path set not to the root dir but to a prefix dir in order to deal with the ["subfolder problem"](https://caddy.community/t/the-subfolder-problem-or-why-cant-i-reverse-proxy-my-app-into-a-subfolder/8575). 

I saw here that a configurable URL Base is in the works, nice! It would be great to easily being able to avoid to give up the web root context for the subdomain entirely to kuma, but having it co-exist alongside other services under the same subdomain. 

I found the nginx config which is great but also rather extensive and possibly prone to breaking (by becoming obsolete quickly as the app changes or potentially from clashing with other services running on the same subdomain). I will try that solution/workaround out but it would be so nice if the app could be deployed in a subdir by setting the URL Base. 

With the contribution from https://github.com/louislam/uptime-kuma/pull/1092#issuecomment-2475806437 getting configurable URL Base seems to progress nicely, but there are concerns about the need to avoid working with a "fat image" with too many dev deps bundled. 

I'd therefore just like to share that I have had good experiences from using multi stage builds (see https://v2.vuejs.org/v2/cookbook/dockerize-vuejs-app.html?redirect=true#Real-World-Example) which essentially makes it possible to have a build with a full dev environment from a base layer and in a subsequent step (in the same Containerfile or Dockerfile) use the outputs in a slimmed down production image which I believe could alleviate the concerns raised.

So just a suggestion that maybe using a multi stage Containerfile/Dockerfile may allow to achive the equivalence of being able to "run a npm run build after Docker pull process" without significant disruption of the existing procedures and without introducing new tooling?

**Comment by CommanderStorm** [comment]
This PR is a bit stuck.
I don't think there is a good solution to the subdirectory problem unless we change how we deliver the software.

I would accept a new PR which fixes most of the issues, but does not have us execute `npm run build` at runtime.
